sdk/containerregistry/azure-containers-containerregistry/src/main/java/com/azure/containers/containerregistry/specialized/ContainerRegistryBlobAsyncClient.java

```diff
@@ -221,28 +214,27 @@ public Mono<UploadBlobResult> uploadBlob(BinaryData data) {
      */
     @ServiceMethod(returns = ReturnType.SINGLE)
     public Mono<Response<UploadBlobResult>> uploadBlobWithResponse(BinaryData data) {
-        return withContext(context -> this.uploadBlobWithResponse(data.toByteBuffer(), context));
+        return withContext(context -> this.uploadBlobWithResponse(data, context));
     }
 
-    private Mono<Response<UploadBlobResult>> uploadBlobWithResponse(ByteBuffer data, Context context) {
+    private Mono<Response<UploadBlobResult>> uploadBlobWithResponse(BinaryData data, Context context) {
         if (data == null) {
             return monoError(LOGGER, new NullPointerException("'data' can't be null."));
         }
 
-        String digest = UtilsImpl.computeDigest(data);
-        return this.blobsImpl.startUploadWithResponseAsync(repositoryName, context)
-            .flatMap(startUploadResponse -> this.blobsImpl.uploadChunkWithResponseAsync(trimNextLink(startUploadResponse.getDeserializedHeaders().getLocation()), Flux.just(data), data.remaining(), context))
-            .flatMap(uploadChunkResponse -> this.blobsImpl.completeUploadWithResponseAsync(digest, trimNextLink(uploadChunkResponse.getDeserializedHeaders().getLocation()), (Flux<ByteBuffer>) null, 0L, context))
-            .flatMap(completeUploadResponse -> {
-                Response<UploadBlobResult> res = new ResponseBase<ContainerRegistryBlobsCompleteUploadHeaders, UploadBlobResult>(completeUploadResponse.getRequest(),
-                    completeUploadResponse.getStatusCode(),
-                    completeUploadResponse.getHeaders(),
-                    new UploadBlobResult(completeUploadResponse.getDeserializedHeaders().getDockerContentDigest()),
-                    completeUploadResponse.getDeserializedHeaders());
-
-                return Mono.just(res);
-            }).onErrorMap(UtilsImpl::mapException);
+        String digest = computeDigest(data.toByteBuffer());
+        return blobsImpl.startUploadWithResponseAsync(repositoryName, context)
+                .flatMap(startUploadResponse -> blobsImpl.uploadChunkWithResponseAsync(trimNextLink(startUploadResponse.getDeserializedHeaders().getLocation()), data.toFluxByteBuffer(), data.getLength(), context))
```

Comment:
Does `uploadChunk` have to upload the entire payload in one request?

Reply:
no, it's the next work item on my list, but hope to finish manifest upload/download first